core/tee/fs_dirfile.c

```diff
@@ -0,0 +1,388 @@
+		dirh->ndents = n;
+		*dirh_ret = dirh;
+	} else {
+		tee_fs_dirfile_close(dirh);
+	}
+	return res;
+}
+
+void tee_fs_dirfile_close(struct tee_fs_dirfile_dirh *dirh)
+{
+	if (dirh) {
+		dirh->fops->close(dirh->fh);
+		free(dirh->files);
+		free(dirh);
+	}
+}
+
+TEE_Result tee_fs_dirfile_commit_writes(struct tee_fs_dirfile_dirh *dirh)
+{
+	return dirh->fops->commit_writes(dirh->fh, NULL);
+}
+
+TEE_Result tee_fs_dirfile_get_tmp(struct tee_fs_dirfile_dirh *dirh,
+				  struct tee_fs_dirfile_fileh *dfh)
+{
+	TEE_Result res;
+	int i = 0;
+
+	if (dirh->files) {
+		bit_ffc(dirh->files, dirh->nbits, &i);
+		if (i == -1)
+			i = dirh->nbits;
+	}
+
+	res = set_file(dirh, i);
+	if (!res)
+		dfh->file_number = i;
+
+	return res;
+}
+
+TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
+			       const void *oid, size_t oidlen,
+			       struct tee_fs_dirfile_fileh *dfh)
+{
+	TEE_Result res;
+	struct dirfile_entry dent;
+	int n;
+	int first_free = -1;
+
+	for (n = 0;; n++) {
+		res = read_dent(dirh, n, &dent);
+		if (res == TEE_ERROR_ITEM_NOT_FOUND && !oidlen) {
+			memset(&dent, 0, sizeof(dent));
+			if (first_free != -1)
+				n = first_free;
+			break;
+		}
+		if (res)
+			return res;
+
+		if (!dent.oidlen && first_free == -1)
+			first_free = n;
+		if (dent.oidlen != oidlen)
+			continue;
+
+		assert(!oidlen || !dent.oidlen ||
+		       test_file(dirh, dent.file_number));
+
+		if (!memcmp(&dent.uuid, dirh->uuid, sizeof(dent.uuid)) &&
+		    !memcmp(&dent.oid, oid, oidlen))
+			break;
+	}
+
+	if (dfh) {
+		dfh->idx = n;
+		dfh->file_number = dent.file_number;
+		memcpy(dfh->hash, dent.hash, sizeof(dent.hash));
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result tee_fs_dirfile_fileh_to_fname(const struct tee_fs_dirfile_fileh *dfh,
+					 char *fname, size_t *fnlen)
+{
+	int r;
+	size_t l = *fnlen;
+
+	if (dfh)
+		r = snprintf(fname, l, "%" PRIx32, dfh->file_number);
+	else
+		r = snprintf(fname, l, "dirf.db");
+
+	if (r < 0)
+		return TEE_ERROR_GENERIC;
+
+	*fnlen = r + 1;
+	if ((size_t)r >= l)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result tee_fs_dirfile_rename(struct tee_fs_dirfile_dirh *dirh,
+				 struct tee_fs_dirfile_fileh *dfh,
+				 const void *oid, size_t oidlen)
+{
+	TEE_Result res;
+	struct dirfile_entry dent;
+
+	if (!oidlen || oidlen > sizeof(dent.oid))
+		return TEE_ERROR_BAD_PARAMETERS;
+	memset(&dent, 0, sizeof(dent));
+	dent.uuid = *dirh->uuid;
+	memcpy(dent.oid, oid, oidlen);
+	dent.oidlen = oidlen;
+	memcpy(dent.hash, dfh->hash, sizeof(dent.hash));
+	dent.file_number = dfh->file_number;
+
+	if (dfh->idx < 0) {
+		struct tee_fs_dirfile_fileh dfh2;
+
+		res = tee_fs_dirfile_find(dirh, oid, oidlen, &dfh2);
+		if (res) {
+			if (res == TEE_ERROR_ITEM_NOT_FOUND)
+				res = tee_fs_dirfile_find(dirh, NULL, 0, &dfh2);
+			if (res)
+				return res;
+		}
+		dfh->idx = dfh2.idx;
+	}
+
+	return write_dent(dirh, dfh->idx, &dent);
+}
+
+TEE_Result tee_fs_dirfile_remove(struct tee_fs_dirfile_dirh *dirh,
```

Comment:
There's much opportunity for optimizations. I didn't include anything like that in this patch to keep things as simple as possible. It will for sure be interesting to have a `tee_fs_dirfile_compact()` or similar. We could keep a `bitstr_t` over used entries to avoid searching for free entries, we could avoid closing this file all the time etc.